### PR TITLE
Add new nft standard

### PR DESCRIPTION
**Pull request by MedovTimur**

### CODE REVIEW

File: contracts/nft/io/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    Burn { token_id: TokenId },
+    Transfer { to: ActorId, token_id: TokenId },
+    Approve { to: ActorId, token_id: TokenId },
+    Owner { token_id: TokenId },
+    IsApproved { to: ActorId, token_id: TokenId },
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum NftEvent {
+    Mint {
```
Comment (LouiseMedova) [inline contracts/nft/io/src/lib.rs:48]: Minted, Burnt, etc ..

File: contracts/nft/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    msg,
+    prelude::*,
+    ActorId,
+};
+use nft_io::*;
+
+#[derive(Debug, Default)]
+pub struct Nft {
+    pub owner_by_id: HashMap<TokenId, ActorId>,
+    pub token_approvals: HashMap<TokenId, HashSet<ActorId>>,
```
Comment (LouiseMedova) [inline contracts/nft/src/lib.rs:14]: I think it's better to make only one approved address

File: contracts/nft/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    pub token_metadata_by_id: HashMap<TokenId, TokenMetadata>,
+    pub tokens_for_owner: HashMap<ActorId, Vec<TokenId>>,
```
Comment (LouiseMedova) [inline contracts/nft/src/lib.rs:16]:
```suggestion
    pub tokens_for_owner: HashMap<ActorId, HashSet<TokenId>>,
```

File: contracts/nft/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+    };
+    NFT = Some(nft);
+}
+
+impl Nft {
+    /// Mint a new nft using `TokenMetadata`
+    fn mint(&mut self, token_metadata: TokenMetadata) -> NftEvent {
+        let msg_src = msg::source();
+        self.check_config();
+        self.check_zero_address(&msg_src);
```
Comment (LouiseMedova) [inline contracts/nft/src/lib.rs:43]: `msg::source` can't be zero address

File: contracts/nft/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        config: init.config,
+        owner: msg::source(),
+        ..Default::default()
+    };
+    NFT = Some(nft);
+}
+
+impl Nft {
+    /// Mint a new nft using `TokenMetadata`
+    fn mint(&mut self, token_metadata: TokenMetadata) -> NftEvent {
```
Comment (LouiseMedova) [inline contracts/nft/src/lib.rs:40]: In general, in the standard it is common to indicate at which address the mint is made:
```suggestion
    fn mint(&mut self, to: &ActorId, token_metadata: TokenMetadata) -> NftEvent {
```

File: contracts/nft/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        self.token_metadata_by_id
+            .insert(self.token_id, token_metadata.clone());
+
+        NftEvent::Mint {
+            to: msg_src,
+            token_metadata,
+        }
+    }
+    /// Burn nft by `TokenId`
+    fn burn(&mut self, token_id: TokenId) -> NftEvent {
```
Comment (LouiseMedova) [inline contracts/nft/src/lib.rs:58]: Approves are also must be removed if any

File: contracts/nft/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    /// Check the existence of a approve
+    fn check_approve(&self, token_id: &TokenId) {
+        if self.token_approvals.contains_key(token_id) {
+            panic!("Approve has already been issued");
+        }
+    }
+}
+
+#[no_mangle]
+unsafe extern fn handle() {
```
Comment (shamilsan) [inline contracts/nft/src/lib.rs:214]: Why `unsafe` function? Please replace to unsafe blocks.
```suggestion
extern fn handle() {
```

File: contracts/nft/io/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    },
+    Transfer {
+        to: ActorId,
+        token_id: TokenId,
+    },
+    Approve {
+        to: ActorId,
+        token_id: TokenId,
+    },
+    Owner {
```
Comment (shamilsan) [inline contracts/nft/io/src/lib.rs:53]: I'd add a verb to the action. `GetOwner` or `GetOwnerOf`.
```suggestion
    GetOwner {
```

File: contracts/nft/io/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        token_id: TokenId,
+    },
+    IsApproved {
```
Comment (shamilsan) [inline contracts/nft/io/src/lib.rs:56]: Also, a verb:
```suggestion
    CheckIfApproved {
```

File: contracts/nft/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+            token_metadata_by_id,
+            tokens_for_owner,
+            token_id,
+            owner,
+            collection,
+            config,
+        } = value;
+
+        let owner_by_id = owner_by_id
+            .iter()
```
Comment (shamilsan) [inline contracts/nft/src/lib.rs:292]: Again, why not `into_iter`?
```suggestion
            .into_iter()
```

**Review by shamilsan — APPROVED** [review]

**Pull request merged**